rclrs/src/lib.rs

```diff
@@ -119,6 +124,81 @@ pub fn create_node_builder(context: &Context, node_name: &str) -> NodeBuilder {
     Node::builder(context, node_name)
 }
 
+/// Extract non-ROS arguments from program's input arguments.
+///
+/// `args` is expected to be input arguments of the program (passed via [`env::args()`])
+/// According to rcl documentation ROS arguments are between `--ros-args` and `--` arguments.
+/// Everything beside is considered as non-ROS and left unparsed. Extracted non-ROS args are
+/// returned in order as they appear (see example below).
+///
+/// # Example
+/// ```
+/// let non_ros_args = rclrs::extract_non_ros_args(vec!["arg1", "--ros-args", "some", "args", "--", "arg2"]);
+/// assert_eq!(non_ros_args.len(), 2);
+/// assert_eq!(non_ros_args[0], "arg1");
+/// assert_eq!(non_ros_args[1], "arg2");
+/// ```
+pub fn extract_non_ros_args(
+    args: impl IntoIterator<Item = String>,
+) -> Result<Vec<String>, RclrsError> {
+    // SAFETY: No preconditions for this function
+    let mut rcl_arguments = unsafe { rcl_get_zero_initialized_arguments() };
+    // SAFETY: No preconditions for this function
+    let allocator = unsafe { rcutils_get_default_allocator() };
+
+    let cstring_args: Vec<CString> = args
+        .into_iter()
+        .map(|arg| {
+            CString::new(arg.as_str()).map_err(|err| RclrsError::StringContainsNul {
+                err,
+                s: arg.clone(),
+            })
+        })
+        .collect::<Result<_, _>>()?;
+    // Vector of pointers into cstring_args
+    let c_args: Vec<*const c_char> = cstring_args.iter().map(|arg| arg.as_ptr()).collect();
+
+    // SAFETY: No preconditions for this function
+    let ret = unsafe {
+        rcl_parse_arguments(
+            c_args.len() as i32,
+            if c_args.is_empty() {
+                std::ptr::null()
+            } else {
+                c_args.as_ptr()
+            },
+            allocator,
+            &mut rcl_arguments,
+        )
+        .ok()
+    };
+
+    if let Err(err) = ret {
+        // SAFETY: No preconditions for this function
+        unsafe {
+            rcl_arguments_fini(&mut rcl_arguments).ok()?;
```

Comment:
I think it's safe to rely on this being done by `rcl` if `rcl_parse_arguments()` fails: https://github.com/ros2/rcl/blob/4eccc3cbe65f5b58981f22efaf612a65731cb2f0/rcl/src/rcl/arguments.c#L670